### DISCUSSION TITLE
[RFC] Forbid empty user identifier

### DISCUSSION THREAD

**Discussion by ajgarlag** [symfony/symfony#57982] [Q&A]
Some built-in authenticators already forbid using an empty user identifier, and I think this is an edge case that should not be supported.

I would add a  `@return non-empty-string` annotation to `UserInterface::getUserIdentifier`. Then, I would add a check in `UserBadge::__construct` that would trigger a deprecation in 7.x and throw a `BadCredentialsException` in 8.x if an empty user identifier is received.

WDYT?
 
<details>
<summary>
Built-in authenticators checking for empty user identifier
</summary>

https://github.com/symfony/symfony/blob/ca1f52806fa61c91d45e56c4981e41c61be3b920/src/Symfony/Component/Security/Http/Authenticator/FormLoginAuthenticator.php#L130-L132

https://github.com/symfony/symfony/blob/ca1f52806fa61c91d45e56c4981e41c61be3b920/src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php#L149-L151

</details>

See #58007

**Comment by mbabker** [comment]
On principle, I agree with the idea here.  But, would this also affect `Symfony\Component\Security\Core\Authentication\Token\TokenInterface::getUserIdentifier()` (given that in practice it mainly just forwards to `Symfony\Component\Security\Core\User\UserInterface::getUserIdentifier()`), and if so, how would cases like the `Symfony\Component\Security\Core\Authentication\Token\NullToken` class (which by design represents an empty token with no user info) be addressed?

**Reply by ajgarlag** [reply]
I see no need to modify `TokenInterface::getUserIdentifier()`.

With my proposed change you can assume that if this method returns an empty string, the token should be a `NullToken` instance.

**Comment by fabpot** [comment]
#58007 has been merged now.

**Comment by ThomasLandauer** [comment]
With this mapping:
```php
/** @var non-empty-string $token */
#[ORM\Column]
#[Assert\NotBlank]
private string $token;
```
PHPStan is now complaining:
> Property App\Entity\User::$... type mapping mismatch: database can contain string but property expects non-empty-string.

Did you find a way around this, or are you just `@phpstan-ignoring` it? :-)

**Reply by mbabker** [reply]
I just ignore that stuff in general because there are cases where the database mapping can't match up with the PHP type mapping (databases don't have a notion of non-empty-string unless I'm really missing something).

```sh
# Fields where the entity and the database cannot have a matching type declaration (typically due to stronger type inference within PHP, i.e. positive-int)
- '#^Property App\\Entity\\[^:]+::\$id type mapping mismatch: database can contain int but property expects int<1, max>\|null\.$#'
- '#^Property App\\Entity\\[^:]+::\$\w+ type mapping mismatch: database can contain int but property expects int<(0|1), max>\.$#'
- '#^Property App\\Entity\\[^:]+::\$\w+ type mapping mismatch: database can contain int\|null but property expects int<(0|1), max>\|null\.$#'
- '#^Property App\\Entity\\[^:]+::\$\w+ type mapping mismatch: database can contain string\|null but property expects non-empty-string\|null\.$#'
- '#^Property App\\Entity\\[^:]+::\$\w+ type mapping mismatch: database can contain string but property expects non-empty-string\.$#'
```

**Reply by ThomasLandauer** [reply]
I suggested a new type for Doctrine that natively supports this: https://github.com/doctrine/orm/issues/11748
